samza-sql/src/main/java/org/apache/samza/sql/translator/SamzaSqlRelMessageJoinFunction.java

```diff
@@ -0,0 +1,115 @@
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.translator;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.commons.lang.Validate;
+import org.apache.samza.operators.KV;
+import org.apache.samza.operators.functions.StreamTableJoinFunction;
+import org.apache.samza.sql.data.SamzaSqlCompositeKey;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.samza.sql.data.SamzaSqlCompositeKey.*;
+
+
+public class SamzaSqlRelMessageJoinFunction implements StreamTableJoinFunction<SamzaSqlCompositeKey, SamzaSqlRelMessage, KV<SamzaSqlCompositeKey, SamzaSqlRelMessage>, SamzaSqlRelMessage> {
+
+  private static final Logger log = LoggerFactory.getLogger(SamzaSqlRelMessageJoinFunction.class);
+
+  private final JoinRelType joinRelType;
+  private final boolean isTablePosOnRight;
+  private final List<Integer> streamFieldIds;
+  // Table field names are used in the outer join when the table record is not found.
+  private final List<String> tableFieldNames;
+  private final List<String> outFieldNames;
+
+  public SamzaSqlRelMessageJoinFunction(JoinRelType joinRelType, boolean isTablePosOnRight,
+      List<Integer> streamFieldIds, List<String> streamFieldNames, List<String> tableFieldNames) {
+    this.joinRelType = joinRelType;
+    this.isTablePosOnRight = isTablePosOnRight;
+    Validate.isTrue((joinRelType.compareTo(JoinRelType.LEFT) == 0 && isTablePosOnRight) ||
+        (joinRelType.compareTo(JoinRelType.RIGHT) == 0 && !isTablePosOnRight) ||
+        joinRelType.compareTo(JoinRelType.INNER) == 0);
+    this.streamFieldIds = streamFieldIds;
+    this.tableFieldNames = tableFieldNames;
+    this.outFieldNames = new ArrayList<>();
+    if (isTablePosOnRight) {
+      outFieldNames.addAll(streamFieldNames);
+    }
+    outFieldNames.addAll(tableFieldNames);
+    if (!isTablePosOnRight) {
+      outFieldNames.addAll(streamFieldNames);
+    }
+  }
+
+  @Override
+  public SamzaSqlRelMessage apply(SamzaSqlRelMessage message, KV<SamzaSqlCompositeKey, SamzaSqlRelMessage> record) {
+
+    if (joinRelType.compareTo(JoinRelType.INNER) == 0 && record == null) {
+      log.debug("Inner Join: Record not found for the message with key: " + getMessageKey(message));
+      return null;
```

Comment:
On innerJoins when you return null, Does this get filtered out? Does the Fluent join function take care of filtering these values out? Can you add a comment if that's the case?